src/ultk/effcomm/information_bottleneck/ba.py

```diff
@@ -0,0 +1,228 @@
+import numpy as np
+
+from typing import Any
+from tqdm import tqdm
+from .tools import (
+    information_cond,
+    random_stochastic_matrix,
+    add_noise_to_stochastic_matrix,
+    compute_lower_bound,
+)
+
+
+##############################################################################
+# Base Rate Distortion class
+##############################################################################
+
+
+class BaseRDOptimizer:
+    def __init__(
+        self,
+        betas: np.ndarray,
+        *args,
+        max_it: int = 1000,
+        eps: float = 1e-15,
+        ignore_converge: bool = False,
+        xhat_size=None,
+        **kwargs,
+    ) -> None:
+        """Base initializer for a Blahut-Arimoto-based optimizer of the Rate Distortion function.
+
+        Args:
+            betas: 1D array, values of beta to search
+
+            max_it: max number of iterations
+
+            args: propagated to `self.beta_iterate` as *kwargs
+
+            eps: accuracy required by the algorithm: the algorithm stops if there is no change in distortion value of more than `eps` between consecutive iterations
+
+            ignore_converge: whether to run the optimization until `max_it`, ignoring the stopping criterion specified by `eps`.
+
+            xhat_size: the size of the output alphabet. The resulting encoders will be of shape (x, xhat)
+
+            kwargs: propagated to `self.beta_iterate` as **kwargs
+        """
+        self.betas = betas
+        self.max_it = max_it
+        self.eps = eps
+        self.ignore_converge = ignore_converge
+
+        self.init_args = args
+        self.init_kwargs = kwargs
+
+        self.ln_px = None  # shape `(x)`
+        self.ln_qxhat_x = None  # shape `(x, xhat)`
+        self.dist_mat = None  # shape `(x, xhat)`
+
+        self.xhat_size = xhat_size
+        # if xhat_size is None:
+        # self.xhat_size = len(self.ln_px)
+
+        self.result = None  # namedtuple
+        self.results: list[Any] = []  # list of namedtuples
```

Comment:
Why not `list[namedtuple]` then?  And same for above.

Reply:
Lol good catch